src/NuGet.Core/NuGet.CommandLine.XPlat/Commands/PackageReferenceCommands/ListPackage/ListPackageCommandRunner.cs

```diff
@@ -0,0 +1,337 @@
+                            else
+                            {
+                                //Handle outdated
+                                if (listPackageArgs.IncludeOutdated)
+                                {
+                                    await AddLatestVersionsAsync(packages, listPackageArgs);
+                                }
+
+                                //Printing packages of a single project and keeping track if
+                                //an auto-referenced package was printed
+                                ProjectPackagesPrintUtility.PrintPackages(packages, projectName, listPackageArgs.IncludeTransitive, listPackageArgs.IncludeOutdated, out var autoRefFoundWithinProject);
+                                autoReferenceFound = autoReferenceFound || autoRefFoundWithinProject;
+                            }
+                        }
+
+                    }
+                    else
+                    {
+                        Console.WriteLine(string.Format(Strings.ListPkg_ErrorReadingAssetsFile, assetsPath));
+                    }
+                    //Unload project
+                    ProjectCollection.GlobalProjectCollection.UnloadProject(project);
+                }
+            }
+
+            //If any auto-references were found, a line is printed
+            //explaining what (A) means
+            if (autoReferenceFound)
+            {
+                Console.WriteLine(Strings.ListPkg_AutoReferenceDescription);
+            }
+        }
+
+        /// <summary>
+        /// Fetches the latest versions for all of the packages that are
+        /// to be listed
+        /// </summary>
+        /// <param name="packages">The packages found in a project</param>
+        /// <param name="listPackageArgs">List args for the token and source provider</param>
+        /// <returns>A data structure like packages, but includes the latest versions</returns>
+        private async Task AddLatestVersionsAsync(
+            IEnumerable<FrameworkPackages> packages, ListPackageArgs listPackageArgs)
+        {
+            var providers = Repository.Provider.GetCoreV3();
+
+            //Handling concurrency and throttling variables
+            var maxTasks = Environment.ProcessorCount;
+            var contactSourcesRunningTasks = new List<Task>();
+            var latestVersionsRequests = new List<Task>();
+
+            //Unique Dictionary for packages and list of latest versions to handle different sources
+            var packagesVersionsDict = new Dictionary<InstalledPackageReference, IList<NuGetVersion>>(new InstalledPackageReferenceIdComparer());
+            AddPackagesToDict(packages, packagesVersionsDict);
+
+            //Prepare requests for each of the packages
+            foreach (var package in packagesVersionsDict)
+            {
+                PrepareLatestVersionsRequests(package.Key, listPackageArgs, providers, latestVersionsRequests, packagesVersionsDict);
```

Comment:
style: `PrepareLatestVersionsRequests` should return `IList<Task>` instead of taking a list and updating it. In current code, it's hard to understand which argument it's going to update.